### DISCUSSION TITLE
[ZK Stack] - How to upgrade system contract?

### DISCUSSION THREAD

**Discussion by SuccinctPaul** [zkSync-Community-Hub/zksync-developers#633] [ZK Stack] [answered]
### Module Affected

- [X] System contracts (era-system-contracts)
- [X] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [ ] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

1.78.0-nightly

### Operating System

- [X] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

1. how to upgrade system contract? 
2. how to upgrade bridge contract? 
3. besides above, what else need to do?

please provide step by step command tutorial to show how to do above.

### Expected Behavior

please provide step by step command tutorial to show how to do upgrade zkstack.

### Current Behavior

No idea.

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by alhenaai** [comment]
### How to Upgrade System and Bridge Contracts in ZKsync Era

#### Upgrading System Contracts
1. **Prepare the Upgrade**:
   - Ensure you have the latest code from the main branch of the [era-contracts repository](https://github.com/matter-labs/era-contracts).
   - Compile the new system contracts that you wish to upgrade.

2. **Initiate the Upgrade**:
   - System upgrades are managed from Layer 1 (L1). You need to create a proposal for the upgrade, which includes the new contract bytecode.
   - Use the governance mechanism to approve the upgrade. This typically involves a multisig or a DAO vote depending on the governance structure.

3. **Execute the Upgrade**:
   - Once approved, execute the upgrade transaction on L1. This will update the system contracts on L2.
   - Verify the upgrade by interacting with the contracts or using the block explorer to ensure the new contracts are active.

#### Upgrading Bridge Contracts
1. **Prepare the Upgrade**:
   - Similar to system contracts, start by preparing the new bridge contract code. Ensure it is tested and ready for deployment.
   - Deploy the new bridge contract facets on L1.

2. **Hash and Propose the Upgrade**:
   - Hash the `diamondCut` and `initCalldata` with the proposed upgrade and save the hash on L1.
   - Propose the upgrade through the governance system.

3. **Execute and Coordinate the Upgrade**:
   - Once approved, coordinate the upgrade across all chains if using multiple ZK chains. This is crucial for maintaining consistency.
   - Execute the upgrade transaction, applying the new bridge contract.

#### Additional Steps
- **Audit and Testing**: Before deploying any upgrades, ensure they are thoroughly audited and tested in a test environment or using an in-memory node.
- **Monitoring and Logging**: After the upgrade, monitor the system for any unexpected behavior. Set up logging to capture important events related to the upgrade.

For a more detailed guide, refer to the [System Upgrades documentation](https://docs.zksync.io/zk-stack/concepts/system-upgrades) and the [Shared Bridges documentation](https://docs.zksync.io/zk-stack/components/shared-bridges#upgrade-mechanism).

This step-by-step guide should help you understand the process of upgrading system and bridge contracts within the ZKsync Era framework.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by bxpana** [reply]
@SuccinctPaul checking to see if you still need support on this

**Reply by SuccinctPaul** [reply]
Yes. need help. 
Please provide a step by step doc.

**Reply by bxpana** [reply]
Will work on getting more info for you on this.

**Reply by bxpana** [reply]
BTW are you from the Caldera team?

**Reply by SuccinctPaul** [reply]
LOL. Just is an enthusiastic developer.

**Reply by bxpana** [reply]
Nice, love to see it 😁

**Reply by bxpana** [reply, marked as answer]
From what I got from the team it's only possible to upgrade the system contract through a protocol upgrade. What upgrades were you trying to make?
